File: ml/src/reason_features.py
"""
The feature vector the phone can compute.

The Why-is-my-Baby-Crying pipeline uses 466 librosa features. Most of its
weight sits in `mel_spec`, `mfcc` and the MFCC deltas -- chroma contributes
nothing to the top forty and tonnetz contributes 0.004 -- and the expensive
ones to reimplement are exactly the ones that do not matter.

So this keeps the families that carry the model and computes them from the
frontend that already has a bit-checked Kotlin twin, rather than from librosa.
That matters more than it sounds: the log-mel transform exists in Python for
training and in Kotlin for inference, and a golden fixture pins them together.
Deriving MFCCs from that same log-mel means the whole 456-dimensional vector
inherits the guarantee, instead of introducing a second chance for the two
languages to disagree about a windowing convention.

Everything else about the pipeline -- the augmentation, SMOTE, the forest and
its hyperparameters -- is unchanged from the original.

    mel_spec   64 bands x {mean, std}                        128
    mfcc       40 coefficients x {mean, std, skew, kurt}     160
    mfcc_d     40 x {mean, std}                               80
    mfcc_d2    40 x {mean, std}                               80
    zcr        {mean, std, max, min}                           4
    rms        {mean, std, max, min}                           4
                                                            ---
                                                            456
"""

import numpy as np
from scipy.stats import kurtosis, skew

from audio import logmel_numpy
from config import HOP_LENGTH, N_MELS, WIN_LENGTH

N_MFCC = 40


def dct2_ortho(x: np.ndarray, n_out: int) -> np.ndarray:
    """
    Orthonormal DCT-II along the last axis, written out rather than imported.

    scipy's version is identical, but this is the definition the Kotlin side
    implements, and having the two sit next to each other in the same form is
    what stops them drifting.
    """
    n = x.shape[-1]
    k = np.arange(n_out)[:, None]
    i = np.arange(n)[None, :]
    basis = np.cos(np.pi * (2 * i + 1) * k / (2 * n))
    scale = np.full((n_out, 1), np.sqrt(2.0 / n))
    scale[0, 0] = np.sqrt(1.0 / n)
    return (basis * scale) @ x[..., None].squeeze(-1) if x.ndim == 1 else \
        x @ (basis * scale).T


def delta(x: np.ndarray) -> np.ndarray:
    """
    Centred first difference along time, edges replicated.

    librosa uses a Savitzky-Golay fit over nine frames. This is the simpler
    definition on purpose: it is one line in both languages and has no window
    or padding convention to get wrong, and for a statistic that is then
    reduced to a mean and a standard deviation the difference does not survive
    the reduction.
    """
    if x.shape[0] < 3:
        return np.zeros_like(x)
    out = np.empty_like(x)
    out[1:-1] = (x[2:] - x[:-2]) * 0.5
    out[0] = out[1]
    out[-1] = out[-2]
    return out


def frame_stats(wave: np.ndarray):
    """Zero-crossing rate and RMS, framed like the spectrogram."""
    wave = np.asarray(wave, dtype=np.float32)
    if wave.size < WIN_LENGTH:
        wave = np.pad(wave, (0, WIN_LENGTH - wave.size))
    frames = 1 + (wave.size - WIN_LENGTH) // HOP_LENGTH
    idx = np.arange(WIN_LENGTH)[None, :] + HOP_LENGTH * np.arange(frames)[:, None]
    framed = wave[idx]

    signs = np.signbit(framed)
    zcr = np.mean(signs[:, 1:] != signs[:, :-1], axis=1)
    rms = np.sqrt(np.mean(framed.astype(np.float64) ** 2, axis=1))
    return zcr.astype(np.float64), rms


def _stats(column: np.ndarray, which):
    out = []
    for stat in which:
        if stat == "mean":
            out.append(float(np.mean(column)))
        elif stat == "std":
            out.append(float(np.std(column)))
        elif stat == "skew":
            out.append(float(skew(column)) if np.std(column) > 1e-12 else 0.0)
        elif stat == "kurt":
            out.append(float(kurtosis(column)) if np.std(column) > 1e-12 else -3.0)
        elif stat == "max":
            out.append(float(np.max(column)))
        elif stat == "min":
            out.append(float(np.min(column)))
    return out


def feature_names():
    names = []
    for i in range(N_MELS):
        names += [f"mel_{i}_mean", f"mel_{i}_std"]
    for i in range(N_MFCC):
        names += [f"mfcc_{i}_mean", f"mfcc_{i}_std",
                  f"mfcc_{i}_skew", f"mfcc_{i}_kurt"]
    for i in range(N_MFCC):
        names += [f"mfccd_{i}_mean", f"mfccd_{i}_std"]
    for i in range(N_MFCC):
        names += [f"mfccdd_{i}_mean", f"mfccdd_{i}_std"]
    names += ["zcr_mean", "zcr_std", "zcr_max", "zcr_min"]
    names += ["rms_mean", "rms_std", "rms_max", "rms_min"]
    return names


def extract(wave: np.ndarray) -> np.ndarray:
    """float32 mono at SAMPLE_RATE -> (408,) float32."""
    mel = logmel_numpy(wave)                      # (frames, 40), natural log
    mfcc = dct2_ortho(mel, N_MFCC)                # (frames, 40)
    d1 = delta(mfcc)
    d2 = delta(d1)
    zcr, rms = frame_stats(wave)

    out = []
    for i in range(N_MELS):
        out += _stats(mel[:, i], ("mean", "std"))
    for i in range(N_MFCC):
        out += _stats(mfcc[:, i], ("mean", "std", "skew", "kurt"))
    for i in range(N_MFCC):
        out += _stats(d1[:, i], ("mean", "std"))
    for i in range(N_MFCC):
        out += _stats(d2[:, i], ("mean", "std"))
    out += _stats(zcr, ("mean", "std", "max", "min"))
    out += _stats(rms, ("mean", "std", "max", "min"))

    return np.asarray(out, dtype=np.float32)

File: android/app/src/main/java/com/nila/audio/ReasonFeatures.kt
package com.nila.audio

import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sqrt

/**
 * The 456 numbers the cause classifier is trained on.
 *
 * This is the Kotlin half of `ml/src/reason_features.py`, and the two are
 * pinned together by a golden fixture the way [LogMelFrontend] already is --
 * everything here is derived from that frontend, so the whole vector inherits
 * a transform that is already known to agree across the two languages rather
 * than introducing a second chance to disagree about a windowing convention.
 *
 *     mel      64 bands  x {mean, std}                       128
 *     mfcc     40 coeffs x {mean, std, skew, kurtosis}        160
 *     mfcc'    40 x {mean, std}                                80
 *     mfcc''   40 x {mean, std}                                80
 *     zcr      {mean, std, max, min}                            4
 *     rms      {mean, std, max, min}                            4
 */
object ReasonFeatures {

    const val N_MFCC = 40
    const val SIZE = 2 * LogMelFrontend.N_MELS + 4 * N_MFCC + 2 * N_MFCC +
        2 * N_MFCC + 4 + 4

    /**
     * Orthonormal DCT-II basis, built once.
     *
     * 40 x 64 doubles. Rebuilding it per call would dominate the cost of the
     * whole extraction, which runs on every classified window.
     */
    private val dctBasis: Array<DoubleArray> by lazy {
        val n = LogMelFrontend.N_MELS
        Array(N_MFCC) { k ->
            val scale = if (k == 0) sqrt(1.0 / n) else sqrt(2.0 / n)
            DoubleArray(n) { i -> cos(PI * (2 * i + 1) * k / (2.0 * n)) * scale }
        }
    }

    /**
     * @param wave  mono float PCM at [LogMelFrontend.SAMPLE_RATE]
     * @param out   [SIZE] floats, overwritten
     */
    fun extract(wave: FloatArray, offset: Int, length: Int, out: FloatArray) {
        require(out.size >= SIZE) { "need $SIZE floats, got ${out.size}" }

        val frames = LogMelFrontend.frameCount(length)
        if (frames < 1) { out.fill(0f); return }

        val mel = FloatArray(frames * LogMelFrontend.N_MELS)
        LogMelFrontend.logMel(wave, offset, length, mel)

        val mfcc = Array(frames) { t ->
            DoubleArray(N_MFCC) { k ->
                var sum = 0.0
                val base = t * LogMelFrontend.N_MELS
                val row = dctBasis[k]
                for (i in 0 until LogMelFrontend.N_MELS) sum += mel[base + i] * row[i]
                sum
            }
        }
        val d1 = delta(mfcc)
        val d2 = delta(d1)

        var w = 0
        // --- mel bands
        val column = DoubleArray(frames)
        for (band in 0 until LogMelFrontend.N_MELS) {
            for (t in 0 until frames) column[t] = mel[t * LogMelFrontend.N_MELS + band].toDouble()
            out[w++] = mean(column).toFloat()
            out[w++] = std(column).toFloat()
        }
        // --- mfcc and its two derivatives
        for (k in 0 until N_MFCC) {
            for (t in 0 until frames) column[t] = mfcc[t][k]
            val m = mean(column)
            val s = std(column)
            out[w++] = m.toFloat()
            out[w++] = s.toFloat()
            out[w++] = skew(column, m, s).toFloat()
            out[w++] = kurtosis(column, m, s).toFloat()
        }
        for (source in listOf(d1, d2)) {
            for (k in 0 until N_MFCC) {
                for (t in 0 until frames) column[t] = source[t][k]
                out[w++] = mean(column).toFloat()
                out[w++] = std(column).toFloat()
            }
        }

        // --- time-domain, framed exactly like the spectrogram
        val zcr = DoubleArray(frames)
        val rms = DoubleArray(frames)
        for (t in 0 until frames) {
            val start = offset + t * LogMelFrontend.HOP_LENGTH
            var crossings = 0
            var energy = 0.0
            var previous = wave[start] < 0f
            for (i in 0 until LogMelFrontend.WIN_LENGTH) {
                val v = wave[start + i]
                energy += v.toDouble() * v
                val negative = v < 0f
                if (i > 0 && negative != previous) crossings++
                previous = negative
            }
            zcr[t] = crossings.toDouble() / (LogMelFrontend.WIN_LENGTH - 1)
            rms[t] = sqrt(energy / LogMelFrontend.WIN_LENGTH)
        }
        for (series in listOf(zcr, rms)) {
            out[w++] = mean(series).toFloat()
            out[w++] = std(series).toFloat()
            out[w++] = series.max().toFloat()
            out[w++] = series.min().toFloat()
        }
    }

    /** Centred first difference along time, edges replicated. */
    private fun delta(x: Array<DoubleArray>): Array<DoubleArray> {
        val frames = x.size
        val width = if (frames > 0) x[0].size else 0
        val out = Array(frames) { DoubleArray(width) }
        if (frames < 3) return out
        for (t in 1 until frames - 1) {
            for (k in 0 until width) out[t][k] = (x[t + 1][k] - x[t - 1][k]) * 0.5
        }
        for (k in 0 until width) {
            out[0][k] = out[1][k]
            out[frames - 1][k] = out[frames - 2][k]
        }
        return out
    }

    private fun mean(x: DoubleArray): Double {
        var sum = 0.0
        for (v in x) sum += v
        return sum / x.size
    }

    /** Population standard deviation, matching numpy's default. */
    private fun std(x: DoubleArray): Double {
        val m = mean(x)
        var sum = 0.0
        for (v in x) { val d = v - m; sum += d * d }
        return sqrt(sum / x.size)
    }

    /**
     * Fisher-Pearson skewness and excess kurtosis, as scipy computes them.
     *
     * The degenerate cases are pinned to what scipy returns for a constant
     * series -- 0 and -3 -- rather than to NaN, because a NaN reaching the
     * forest turns every comparison against it into a silent left-branch.
     */
    private fun skew(x: DoubleArray, m: Double, s: Double): Double {
        if (s < 1e-12) return 0.0
        var sum = 0.0
        for (v in x) { val d = (v - m) / s; sum += d * d * d }
        return sum / x.size
    }

    private fun kurtosis(x: DoubleArray, m: Double, s: Double): Double {
        if (s < 1e-12) return -3.0
        var sum = 0.0
        for (v in x) { val d = (v - m) / s; sum += d * d * d * d }
        return sum / x.size - 3.0
    }
}
